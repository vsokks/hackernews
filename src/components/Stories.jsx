import {useEffect, useState} from "react";
import {getStoriesWithDetail} from "../services/StoryService";
import {Button, Col, Container, Row} from "react-bootstrap";
import Story from "./Story";
import {MAX_STORIES_COUNT} from "../constants";

const Stories = ({type}) => {
    const [pageNumber, setPageNumber] = useState(1);
    const [stories, setStories] = useState([]);
    const [isLoading, setIsLoading] = useState(false);

    const getStories = () => {
        setIsLoading(true);
        getStoriesWithDetail(type, pageNumber).then((items) => {
            (items.length > 0) && setStories(stories.concat(items));
            setIsLoading(false);
        });
    }

    useEffect(() => {
        console.log("Executing use effect");
        getStories();
    }, [pageNumber]);

    return (
        <Container fluid>
            {isLoading && stories.length == 0 && <h5>Loading... </h5>}
            {stories.map((story) => <Story story={story}/>)}
            {stories.length > 0
                && stories.length < MAX_STORIES_COUNT
                && <Container fluid>
                        <Row>
                            <Col xs={{ span: 4, offset: 4 }}>
                                <Button onClick={ () => setPageNumber(pageNumber+1)} style={{width: "100%"}}>
                                    {isLoading && "Loading More Stories..."}
                                    {!isLoading && "More Stories"}
                                </Button>
                            </Col>
                        </Row>
                    </Container>
            }
        </Container>
    );
}

export default Stories;