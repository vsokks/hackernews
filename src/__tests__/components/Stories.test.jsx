import React, {useState as useStateMock} from 'react';
import renderer from 'react-test-renderer';
import Stories from "../../components/Stories";

jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useState: jest.fn(),
}));

const setState = jest.fn();
beforeEach(() => {
    useStateMock.mockImplementation(init => [init, setState]);
});

test('Test Stories', () => {
    /*Still Pending, problem in setting mock state using react-test-renderer lib*/
    const component = renderer.create(<Stories type='top' />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});