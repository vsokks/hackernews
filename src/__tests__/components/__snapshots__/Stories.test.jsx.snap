// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Stories 1`] = `
<div
  className="container-fluid"
/>
`;
